core/network/subnet.go

```diff
@@ -155,5 +157,94 @@ func FindSubnetIDsForAvailabilityZone(zoneName string, subnetsToZones map[Id][]s
 		return nil, errors.NotFoundf("subnets in AZ %q", zoneName)
 	}
 
-	return matchingSubnetIDs.SortedValues(), nil
+	sorted := make([]Id, matchingSubnetIDs.Size())
+	for k, v := range matchingSubnetIDs.SortedValues() {
+		sorted[k] = Id(v)
+	}
+
+	return sorted, nil
+}
+
+// SubnetSet represents the classic "set" data structure, and contains Id.
+// SubnetSet is used as a typed version to prevent string -> Id -> string
+// conversion when using set.Strings
+type SubnetSet map[Id]struct{}
+
+// MakeSubnetSet creates and initializes a SubnetSet and populates it with
+// initial values as specified in the parameters.
+func MakeSubnetSet(values ...Id) SubnetSet {
+	set := make(map[Id]struct{}, len(values))
+	for _, id := range values {
+		set[id] = struct{}{}
+	}
+	return set
+}
+
+// Add puts a value into the set.
+func (s SubnetSet) Add(value Id) {
+	s[value] = struct{}{}
+}
+
+// Size returns the number of elements in the set.
+func (s SubnetSet) Size() int {
+	return len(s)
+}
+
+// IsEmpty is true for empty or uninitialized sets.
+func (s SubnetSet) IsEmpty() bool {
+	return len(s) == 0
+}
+
+// Contains returns true if the value is in the set, and false otherwise.
+func (s SubnetSet) Contains(id Id) bool {
+	_, exists := s[id]
+	return exists
+}
+
+// Values returns an unordered slice containing all the values in the set.
+func (s SubnetSet) Values() []Id {
+	result := make([]Id, len(s))
+	i := 0
+	for key := range s {
+		result[i] = key
+		i++
+	}
+	return result
+}
+
+// SortedValues returns an ordered slice containing all the values in the set.
+func (s SubnetSet) SortedValues() []Id {
+	values := s.Values()
+	sort.Slice(values, func(i, j int) bool {
+		return values[i] < values[j]
+	})
+	return values
+}
+
+// FanType describes a network fan type.
+type FanType string
```

Comment:
Isn't this overcooked?

It's just a `const` used in fan network ID strings. I can't see it used in any other form. I don't think we need an exported function that accepts a `FanType`.

Reply:
K, sure, I'll rewind that.

Reply:
Having thought about it, I'm actually unsure, I know we don't probably have a use for it, but the lack of types esp. in-network (consider `network.Id` proliferation) has become the new `string, string, string`.

The type is actually doing us a dis-service here and we should be creating more feature-rich types.

Consider the following function type:

```go
type X interface{
    Create(string, corenetwork.Id, corenetwork.SubnetId) (corenetwork.PortId, error)
}
```
I would rather have that, than:

```go
type X interface {
    Create(string, corenetwork.Id, corenetwork.Id) (string, error)
}
```

I know this is egging the pudding a bit, but actually the types become very explanatory about what's required and what to expect.

The fact that most functions in providers are `func(string, string, map[string]string, string) (string, string, error)` makes it harder to understand how to work with providers.

If we're worried about raw performance then `type FanType = string` is a possibility I believe.

Reply:
I agree with the typed IDs thing, but I can't see `FanType` ever being an argument to a method.

Reply:
OK, we're aligned.